File: src/ObjectManager/Object.Tes/FilePacks/Records/300-TES3.Plugin info.cs
using OA.Core;
using System;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class TES3Record : Record
    {
        public struct HEDRField
        {
            public float Version;
            public uint FileType;
            public string CompanyName;
            public string FileDescription;
            public uint NumRecords;

            public HEDRField(UnityBinaryReader r, uint dataSize)
            {
                Version = r.ReadLESingle();
                FileType = r.ReadLEUInt32();
                CompanyName = r.ReadASCIIString(32, ASCIIFormat.ZeroPadded);
                FileDescription = r.ReadASCIIString(256, ASCIIFormat.ZeroPadded);
                NumRecords = r.ReadLEUInt32();
            }
        }

        public HEDRField HEDR;
        public List<STRVField> MASTs;
        public List<INTVField> DATAs;

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, uint dataSize)
        {
            switch (type)
            {
                case "HEDR": HEDR = new HEDRField(r, dataSize); return true;
                case "MAST": if (MASTs == null) MASTs = new List<STRVField>(); MASTs.Add(new STRVField(r, dataSize)); return true;
                case "DATA": if (DATAs == null) DATAs = new List<INTVField>(); DATAs.Add(new INTVField(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/300-TES3.Plugin info.swift
//
//  TES3Record.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class TES3Record: Record {
    public struct HEDRField: CustomStringConvertible {
        public var description: String { return "TES3" }
        public let version: Float
        public let fileType: UInt32
        public let companyName: String
        public let fileDescription: String
        public let numRecords: UInt32

        init(_ r: BinaryReader, _ dataSize: Int) {
            version = r.readLESingle()
            fileType = r.readLEUInt32()
            companyName = r.readASCIIString(32, format: .zeroPadded)
            fileDescription = r.readASCIIString(256, format: .zeroPadded)
            numRecords = r.readLEUInt32()
        }
    }

    public var HEDR: HEDRField? = nil
    public var MASTs: [STRVField]? = nil
    public var DATAs: [INTVField]? = nil

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        guard format == .TES3 else {
            return false
        }
        switch type {
        case "HEDR": HEDR = HEDRField(r, dataSize)
        case "MAST": if MASTs == nil { MASTs = [STRVField]() }; MASTs!.append(r.readSTRV(dataSize))
        case "DATA": if DATAs == nil { DATAs = [INTVField]() }; DATAs!.append(r.readINTV(dataSize))
        default: return false
        }
        return true
    }
}
